### PR TITLE
fix(api): load external mcp tools once per send with bounded discovery

### PR DESCRIPTION
**Pull request by jan-kubica**
External MCP connector tools were discovered twice per chat send (validation pass and streaming pass, independent and uncached), with connectors loaded sequentially and only a per-HTTP-call 10s bound — one unresponsive third-party server multiplied into tens of seconds on every send for that user.

- Tools are now loaded once per request and reused by both passes, with explicit close-ownership on every exit path (validation failure, early returns, streaming handoff).
- Connectors load concurrently (they are independent per-user integrations; per-connector failure still degrades to "no tools from that connector") under a 20s per-connector aggregate discovery ceiling that also covers auth chaining, not just individual HTTP calls; merge order stays deterministic by row order.
- A validation-tools helper keeps the handler under the cognitive-complexity budget. Chat tools suite green (191 tests); api typecheck clean.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved reliability when connecting to external tools, with safer handling of slow, failing, or timed-out tool discovery.
  * Prevented chat sessions from stalling during external tool validation or streaming.
  * Ensured external tool connections are closed correctly across both normal and error scenarios, avoiding double-close issues.
* **Performance**
  * External tool discovery now runs concurrently where possible, reducing setup time.
* **Tests**
  * Added coverage for timeout and lifecycle edge cases to prevent regressions.

### CODE REVIEW

File: apps/api/src/handlers/chat/tools/external-mcp-tools.ts [hunk 1]
```diff
@@ -129,60 +162,68 @@ const copyServerTools = ({
...
+      },
+    );
   } catch (error) {
     captureError(error, {
       source: "external-mcp-tools",
       connectorSlug: row.slug,
     });
+    return null;
   }
 };
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/handlers/chat/tools/external-mcp-tools.ts:328]: Fixed (47065d85a + 6fef1bfd2): the created client is tracked and closed on every failure path; closeAbandonedClient does a best-effort idempotent close and captures close errors.

File: apps/api/src/handlers/chat/tools/external-mcp-tools.ts [hunk 2]
```diff
@@ -129,60 +162,68 @@ const copyServerTools = ({
...
-    Object.assign(loadedTools, normalized.tools);
-    copyServerTools({ sourceTools, tools: normalized.tools });
-
-    connectors.push({
-      description: row.description,
-      displayName: row.displayName,
-      slug: row.slug,
-      toolNames: normalized.toolNames,
-    });
+        const tools = await loadMcpConnectorTools({ client, row });
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/handlers/chat/tools/external-mcp-tools.ts]: Fixed: the client is now closed on both the throw and timeout paths (see the timeout thread for the immediate-close detail).

File: apps/api/src/handlers/chat/tools/external-mcp-tools.ts [hunk 3]
```diff
@@ -129,60 +164,120 @@ const copyServerTools = ({
...
+  } catch (error) {
+    if (error instanceof TimeoutError) {
+      // `discovery` is still running; wait for it to settle (success or
+      // failure) and close whatever client it created, since the timeout
+      // already discarded its result and no caller will ever see it.
+      void discovery
+        .catch(() => undefined)
+        .then(
+          async () =>
+            await closeAbandonedClient({ client, connectorSlug: row.slug }),
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/handlers/chat/tools/external-mcp-tools.ts:314]: Fixed (6fef1bfd2): confirmed MCPClient.close() synchronously aborts the in-flight fetch and rejects the pending listTools call (traced through the transport's AbortController + Protocol._onclose), so on timeout we now close the known client immediately rather than waiting on the hung discovery promise. A best-effort deferred close still covers the case where the hang was in client creation (client still undefined at timeout); close() is idempotent so there's no double-close. New test pins the immediate close when tools() hangs.

File: apps/api/src/handlers/chat/send-message.ts [hunk 4]
```diff
@@ -287,16 +288,22 @@ const sendMessage = createSafeRootHandler(
...
+    // time right before streaming. `externalMcpToolsHandedOffToStreaming`
+    // below tracks whether ownership of closing these connectors passed
+    // to the streaming try/catch, so every other exit path (validation
+    // failure, any early return between here and streaming) still closes
+    // them exactly once.
+    const externalMcpTools = await loadExternalMcpToolsForUser({
+      nullUnionStrategy: externalMcpNullUnionStrategy,
+      organizationId: session.activeOrganizationId,
+      safeDb,
+      userId: user.id,
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/handlers/chat/send-message.ts]: Good catch (6fef1bfd2): the try/finally now starts immediately after loadExternalMcpToolsForUser (before web-search provider load, tool-set construction, and validation), so a throw from any of those awaited steps closes the MCP clients instead of leaking them. The streaming pass still takes over close-ownership via externalMcpToolsHandedOffToStreaming, so no double-close.

File: apps/api/src/handlers/chat/send-message.ts [hunk 5]
```diff
@@ -287,698 +288,728 @@ const sendMessage = createSafeRootHandler(
...
+    // Loaded once and reused for both the validation and streaming tool
+    // sets (and for the streaming `externalMcpToolSource`) instead of
+    // running connector discovery twice per send: the streaming pass
+    // always needs this set, so load it up front rather than a second
+    // time right before streaming. `externalMcpToolsHandedOffToStreaming`
+    // below tracks whether ownership of closing these connectors passed
+    // to the streaming try/catch, so every other exit path (validation
+    // failure, any early return or throw between here and streaming) still
+    // closes them exactly once.
+    const externalMcpTools = await loadExternalMcpToolsForUser({
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/handlers/chat/send-message.ts]: Fixed (6a9119660): replaced the eager unconditional load with a lazy memoized loader. Validation only triggers discovery when messageNeedsExternalMcpValidation() is true; the streaming pass triggers the same memoized load at the original late point (right before building chatTools), so a message that needs neither — or one that rejects early on malformed parts / abort — never contacts a connector. Both load and close are memoized: never-loaded closes to a no-op, loaded-once closes exactly once, and the streaming handoff ownership is preserved, so no leak and no double-close. Tests cover all four cases.

File: apps/api/src/handlers/chat/tools/external-mcp-tools.ts [hunk 6]
```diff
@@ -129,60 +217,133 @@ const copyServerTools = ({
...
+      await closeAbandonedClient({ client, connectorSlug: row.slug });
+      // Still attach a best-effort close to `discovery`'s eventual
+      // settlement: it covers the case above where no client existed yet
+      // at timeout, and is a no-op otherwise since `MCPClient.close()` is
+      // idempotent (safe even if it ends up closing the same client twice).
+      void discovery
+        .catch(() => undefined)
+        .then(
+          async () =>
+            await closeAbandonedClient({ client, connectorSlug: row.slug }),
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/handlers/chat/tools/external-mcp-tools.ts:314]: This case is bounded, not an unbounded leak. When the aggregate timeout fires while createMcpClientForConnection is still pending, there is no client handle yet to close synchronously — so the deferred discovery.catch(...).then(closeAbandonedClient) is the only possible mechanism, and it does close the late-created client once discovery settles. discovery is itself bounded by the per-call MCP_HTTP_REQUEST_TIMEOUT_MS (~10s) inside mcp-upstream/connections.ts, so the worst case is the client staying open ~10s past the 20s aggregate timeout before being closed, in the rare race where creation finishes after the timeout. There's no earlier point to close it (the handle doesn't exist until createClient resolves). Closing it sooner would require threading an AbortSignal into createMcpClientForConnection to cancel creation itself — a larger change worth its own issue rather than this PR.
